python/uv/private/pin.bzl

```diff
@@ -0,0 +1,184 @@
+        "--generate-hashes",
+        "--emit-index-url",
+        "--no-strip-extras",
+        "--python=$(PYTHON3)",
+    ] + [
+        "$(location {})".format(src)
+        for src in srcs
+    ] + [
+        "--output-file=$(location {})".format(_out),
+    ]
+    if upgrade:
+        args.append("--upgrade")
+    if universal:
+        args.append("--universal")
+    cmd = "$(UV_BIN) pip compile " + " ".join(args)
+
+    # Check if the output file already exists, if yes, first copy it to the
+    # output file location in order to make `uv` not change the requirements if
+    # we are just running the command.
+    if native.glob([out]):
+        cmd = "cp -v $(location {}) $@; {}".format(out, cmd)
+        srcs.append(out)
+
+    native.genrule(
+        name = name + ".uv.out",
+        srcs = srcs,
+        outs = [_out],
+        cmd_bash = cmd,
+        tags = [
+            "local",
+            "manual",
+            "no-cache",
+        ],
+        target_compatible_with = _REQUIREMENTS_TARGET_COMPATIBLE_WITH,
+        toolchains = [
+            Label("//python/uv:current_toolchain"),
+            Label("//python:current_py_toolchain"),
+        ],
+    )
+    if python_version:
+        transitioned_name = "{}.uv.out.{}".format(name, python_version)
+        _versioned(
+            name = transitioned_name,
+            src = _out,
+            python_version = python_version,
+            tags = ["manual"],
+        )
+        _out = transitioned_name
+
+    # Write a script that can be used for updating the in-tree version of the
+    # requirements file
+    write_file(
+        name = name + ".gen",
+        out = name + ".gen.py",
+        content = [
+            "from os import environ",
+            "from pathlib import Path",
+            "from sys import stderr",
+            "",
+            'src = Path(environ["REQUIREMENTS_FILE"])',
+            'dst = Path(environ["BUILD_WORKSPACE_DIRECTORY"]) / "{}" / "{}"'.format(pkg, out),
+            'print(f"Writing requirements contents\\n  from {src.absolute()}\\n  to {dst.absolute()}", file=stderr)',
+            "dst.write_text(src.read_text())",
+            'print("Success!", file=stderr)',
+        ],
+        target_compatible_with = _REQUIREMENTS_TARGET_COMPATIBLE_WITH,
+    )
+
+    py_binary(
+        name = name,
+        srcs = [name + ".gen.py"],
+        main = name + ".gen.py",
+        data = [_out],
+        env = {
+            "REQUIREMENTS_FILE": "$(location {})".format(_out),
+        },
+        tags = ["manual"],
+        target_compatible_with = _REQUIREMENTS_TARGET_COMPATIBLE_WITH,
+    )
+
+def _transition_python_version_impl(_, attr):
+    return {"//python/config_settings:python_version": str(attr.python_version)}
+
+_transition_python_version = transition(
+    implementation = _transition_python_version_impl,
+    inputs = [],
+    outputs = ["//python/config_settings:python_version"],
+)
+
+def _impl(ctx):
```

Comment:
Can you use the code from //python/config_settings:transition.bzl instead? You should be able to just load py_binary from there and set python_version.